tests/codex/blockexchange/engine/testengine.nim

```diff
@@ -344,9 +349,168 @@ asyncchecksuite "NetworkStore engine handlers":
     )
 
     await engine.blocksDeliveryHandler(peerId, blocksDelivery)
-    discard await allFinished(pending)
+    discard await allFinished(pending).wait(100.millis)
     await allFuturesThrowing(cancellations.values().toSeq)
 
+asyncchecksuite "Block Download":
+  var
+    rng: Rng
+    seckey: PrivateKey
+    peerId: PeerId
+    chunker: Chunker
+    wallet: WalletRef
+    blockDiscovery: Discovery
+    peerStore: PeerCtxStore
+    pendingBlocks: PendingBlocksManager
+    network: BlockExcNetwork
+    engine: BlockExcEngine
+    discovery: DiscoveryEngine
+    advertiser: Advertiser
+    peerCtx: BlockExcPeerCtx
+    localStore: BlockStore
+    blocks: seq[Block]
+
+  setup:
+    rng = Rng.instance()
+    chunker = RandomChunker.new(rng, size = 1024'nb, chunkSize = 256'nb)
+
+    while true:
+      let chunk = await chunker.getBytes()
+      if chunk.len <= 0:
+        break
+
+      blocks.add(Block.new(chunk).tryGet())
+
+    seckey = PrivateKey.random(rng[]).tryGet()
+    peerId = PeerId.init(seckey.getPublicKey().tryGet()).tryGet()
+    wallet = WalletRef.example
+    blockDiscovery = Discovery.new()
+    peerStore = PeerCtxStore.new()
+    pendingBlocks = PendingBlocksManager.new()
+
+    localStore = CacheStore.new()
+    network = BlockExcNetwork()
+
+    discovery =
+      DiscoveryEngine.new(localStore, peerStore, network, blockDiscovery, pendingBlocks)
+
+    advertiser = Advertiser.new(localStore, blockDiscovery)
+
+    engine = BlockExcEngine.new(
+      localStore, wallet, network, discovery, advertiser, peerStore, pendingBlocks
+    )
+
+    peerCtx = BlockExcPeerCtx(id: peerId)
+    engine.peers.add(peerCtx)
+
+  test "Should exhaust retries":
+    var
+      retries = 2
```

Comment:
maybe use const for that (we use this number in two places), e.g. `const BlockRetires = 2`?

Reply:
yes, I meant something different - wasn't clear enough I suppose. I meant this:

```nim
const BlockRetires = 2
var
  retries = BlockRetires
  address = BlockAddress.init(blocks[0].cid)

proc sendWantList(
    id: PeerId,
    addresses: seq[BlockAddress],
    priority: int32 = 0,
    cancel: bool = false,
    wantType: WantType = WantType.WantHave,
    full: bool = false,
    sendDontHave: bool = false,
) {.gcsafe, async.} =
  check wantType == WantHave
  check not engine.pendingBlocks.isInFlight(address)
  check engine.pendingBlocks.retries(address) == retries
  retries -= 1

engine.pendingBlocks.blockRetries = BlockRetires
```

This way it is more explicit that those two are related - not a big deal.